### Brief Overview 

- The goal of this assignment is to practice some important data wrangling functionality commonly required in real-world projects.

- Here we will use two datasets:
  - IRS Statistics of Income (SOI) dataset
  - The Medicaid Data per State  


- The final product here is a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  - Which medications account for the bulk of a state's spending   
  - Which drugs are prescribed much more in one state compared to the other states.
etc.



### Instructions on answering the questions

- Some of the questions below require that you only use methods or properties of a `DataFrame` or a `Series`. Therefore, any solution that uses a function that is not a method or property of `DataFrame`s or `Series` will not be accepted, even when the solution yields an appropriate answer to the question.

- For instance, if you are asked to find the number of entries in the DataFrame `tax_data ` using only the DataFrame's methods or properties, then `len(tax_data)` is not an acceptable solution since `len()` is not a `tax_data` method. The statements below are both correct answers:

  - `info()` is a method `tax_data`
  
```python
   
    tax_data.info()
```

  or

  - `shape` is a property of `tax_data`


```python
tax_data.shape
```

- Similarly, if you are asked to find to count the number of unique entries in the `STATEFIPS` column of the `tax_data` DataFrame, then solutions using set() or len() are not acceptable for the following reasons:

- The solution below uses `set()` and `len()`, which are not `tax_data` methods

```python
len(set(tax_data["STATEFIPS"]))
```

- The solution below uses `unique()`, which is a  `tax_data` method, but then counts the number of uniques entries using `len()` which is not `tax_data` method

```python
len(tax_data["STATEFIPS"].unique())
```


- The statement below is an acceptable solution since it uses `nunique()` which is a method of the `Series` generated by indexing on a column of `tax_data` (`tax_data['STATEFIPS'])

```python
tax_data['STATEFIPS'].nunique()
```


- Chaining methods and properties is encouraged if it does not cause ambiguity 

- For instance, to identify whether a value is part of the index, write the following:

```python
tax_data.index.contains(99999)
```

- Rather than:

```python
9999 in tax_data.index
```

- You can only import `pandas` and `numpy`

- If you are not explicitely asked to only use methods or properties of a `DataFrame` or a `Series`, then any solution that does not rely on external products will be accepted.

- __Important__: Provide the exact statement(s) used to answer each question

- Unless otherwise specified, each cell can contain multiple lines of code

* Finally, note that not all the functions necessary for answering the questions below were covered in class. As such, I suggest you use `.SHIFT+TAB` on objects liberally to see which methods and properties are available on objects. If you are unsure what a method does, use `SHIFT+TAB` twice to invoke the `docstring`, or documentation for that function. This is not only a good way to see which functionality can be used to answer the questions below but also a great way to familiarize yourself with the plethora or functionality available through the `pandas` package.

---

In [1]:
import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (tax_data.csv) into a `DataFrame` called `tax_data`. The file is `tax_data.csv` is located in the `data` directory of the assignment folder.

* This dataset was preprocessed but the original one was obtained at the following URL:

        https://www.irs.gov/pub/irs-soi/15zpallagi.csv


In [2]:
## WRITE YOUR CODE HERE 
tax_data = pd.read_csv('data/tax_data.csv')

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`

In [3]:
## WRITE YOUR CODE HERE 
tax_data.head(8)

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2  ...    N10300     A10300   N85530   A85530   N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0      0.0      0.0      0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0      0.0      0.0      0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0      0.0      0.0      0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0      0.0      0.0      0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0    420.0    168.0     60.0   
5   41410.0   160530.0  ...   55300.0  6027793.0  22090.0  39519.0  27550.0   
6     700.0     2160.0  ...     690.0      610.0      0.0      0.0      0.0   
7     610.0     2540.0  ...    1140.0     3019.0      0.0      0.0      0.0   

    A85300   N11901    A11901    N11902     A11902  
0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0  45460.0  128823.0  116440.0   377177.0  
4     31.0  83330.0  421004.0  121570.0   483682.0  
5  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0    120.0      94.0    1290.0     2792.0  
7      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 91 columns]

*  Modify `tax_data` to uppercase all the header name. 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself
  *  You can `tax_data.columns`, which returns a `Series` of the column names.
  
* The resulting column name should look as follows:


```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

This operation is useful for standardizing column names and avoid guessing whether the column header was in upper case, lower case or a mix of both.

In [4]:
tax_data.columns = tax_data.columns.str.upper()

* What is the total number of entries (also called observations) in `tax_data`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [5]:
## WRITE YOUR CODE HERE 
tax_data.shape

(166698, 91)

- If `STATEFIPS` is header title of the first column of the `tax_data` `DataFrame`, what is the title of the 32nd column
  - Your answer can only use `DataFrame` or `Series` methods or properties and should use a single python expression



In [6]:
## WRITE YOUR CODE HERE 
tax_data.columns[31]

'N00900'

* If `STATEFIPS` is the the first column, what is the index of the column name `N10300`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [7]:
## WRITE YOUR CODE HERE 
tax_data.columns.get_loc('N10300')

81

- Print the count of unique zip codes in each state using descending order.
  - Your answer can only use `DataFrame` or `Series` methods or properties

- The result should look like the following ( '...' represents remaining data that is not shown)
```python

        ZIPCODE
STATE	
TX	     9714
NY	     9238
CA	     8908
PA	     8208
IL	     7386
...
```

* The above indicates that there are 9,714 unique zipcodes in TX, 9,238 in NY, etc.


In [8]:
## WRITE YOUR CODE HERE 
states_zips = tax_data.groupby('STATE')['ZIPCODE'].nunique()
states_zips = states_zips.reset_index()
states_zips = states_zips.sort_values(by='ZIPCODE', ascending=False)
states_zips

STATE  ZIPCODE
43    TX     1619
34    NY     1540
4     CA     1485
38    PA     1368
14    IL     1231
35    OH      998
9     FL      920
22    MI      892
24    MO      890
12    IA      827
45    VA      793
23    MN      792
27    NC      725
48    WI      714
15    IN      674
10    GA      667
17    KY      656
16    KS      598
42    TN      590
1     AL      577
31    NJ      547
36    OK      547
49    WV      512
47    WA      497
2     AR      490
29    NE      486
19    MA      482
18    LA      453
20    MD      405
5     CO      395
40    SC      376
25    MS      370
21    ME      370
37    OR      356
3     AZ      295
28    ND      287
41    SD      287
6     CT      264
46    VT      238
30    NH      232
26    MT      225
13    ID      213
32    NM      209
44    UT      186
33    NV      128
50    WY      109
39    RI       71
11    HI       60
8     DE       57
0     AK       56
7     DC       24

- Identify the position of HI in the list of zip code counts per state (questions directly above)
  - Your answer can only use `DataFrame` or `Series` methods or properties

In [9]:
## WRITE YOUR CODE HERE
states_zips.loc[states_zips['STATE'] == 'HI']

STATE  ZIPCODE
11    HI       60

### Identifying and Removing Ambiguous Zip Codes

- Count the number of entries where ZIPCODE is 0, assign your results to a variable named  `nb_invalid_zip`

In [10]:
## WRITE YOUR CODE HERE 
nb_invalid_zip = tax_data['ZIPCODE'].value_counts()[0.0].tolist()
#type(nb_invalid_zip)
print(nb_invalid_zip)

306


* Run the line below to make sure that `nb_invalid_zip` is an integer (`int`)
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`
  
* If `nb_invalid_zip` then change your answer above so that the value returned is effectively a number.

In [11]:
assert(type(nb_invalid_zip) == int)

- Remove from `tax_data` all the lines where the zip code is `0` and save resulting `DataFrame` to a variable named `tax_data_valid_zip`
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [12]:
tax_data_valid_zip = tax_data[tax_data['ZIPCODE'] != 0.0]
tax_data_valid_zip.shape

(166392, 91)

* Run the line below to confirm that the operation worked as expected

  * The assertion below is testing that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [13]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

* How many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`?
  * Your answer can only use `DataFrame` methods and properties
* Assing the count of `NaN` into a variable called nb_missing_values

In [14]:
## WRITE YOUR CODE HERE
nb_missing_values = tax_data_valid_zip.isnull().any(axis=1).sum()
#tax_data_valid_zip.info()
nb_missing_values

139

* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values

* Call the new `DataFrame` `tax_data_valid_zip_cleaned

In [15]:
## WRITE YOUR CODE HERE 
tax_data_valid_zip_cleaned = tax_data_valid_zip.dropna(axis=0)
tax_data_valid_zip_cleaned.shape

(166253, 91)

* Run the line below to confirm that the operation worked as expected. The assertion below is testing that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`
  
* Note that assert will only print an error if the results do not match

In [16]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code

* The function `compute_percentile_zipcode` below computes the percentile income per zip code

* By default percentile=0.5,  i.e., the function computes the median

* Read the code and make sure you understand what it does before moving on to the next question


In [17]:
def compute_percentile_zip(df_zip, percentile=0.5):
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

In [18]:
 tax_data_valid_zip_cleaned = tax_data_valid_zip_cleaned.reset_index(drop=True)
tax_data_valid_zip_cleaned

STATEFIPS STATE  ZIPCODE  AGI_STUB      N1   MARS1   MARS2  MARS4  \
0             1.0    AL  35004.0       1.0  1490.0   970.0   230.0  280.0   
1             1.0    AL  35004.0       2.0  1350.0   630.0   360.0  300.0   
2             1.0    AL  35004.0       3.0   970.0   310.0   490.0  140.0   
3             1.0    AL  35004.0       4.0   620.0   110.0   470.0   30.0   
4             1.0    AL  35004.0       5.0   620.0    40.0   560.0   20.0   
...           ...   ...      ...       ...     ...     ...     ...    ...   
166248       56.0    WY  99999.0       2.0  3680.0  2000.0  1080.0  490.0   
166249       56.0    WY  99999.0       3.0  2630.0  1100.0  1240.0  250.0   
166250       56.0    WY  99999.0       4.0  2020.0   590.0  1310.0  100.0   
166251       56.0    WY  99999.0       5.0  2670.0   430.0  2170.0   90.0   
166252       56.0    WY  99999.0       6.0   760.0   140.0   650.0    0.0   

          PREP      N2  ...  N10300    A10300  N85530  A85530  N85300  A85300  \
0        700.0  2160.0  ...   690.0     610.0     0.0     0.0     0.0     0.0   
1        610.0  2540.0  ...  1140.0    3019.0     0.0     0.0     0.0     0.0   
2        450.0  2160.0  ...   930.0    5009.0     0.0     0.0     0.0     0.0   
3        300.0  1610.0  ...   620.0    5190.0     0.0     0.0     0.0     0.0   
4        320.0  1770.0  ...   620.0   10129.0     0.0     0.0     0.0     0.0   
...        ...     ...  ...     ...       ...     ...     ...     ...     ...   
166248  2020.0  6660.0  ...  3220.0    9235.0     0.0     0.0     0.0     0.0   
166249  1610.0  5440.0  ...  2530.0   15579.0     0.0     0.0     0.0     0.0   
166250  1330.0  4780.0  ...  1990.0   19278.0     0.0     0.0     0.0     0.0   
166251  1850.0  6930.0  ...  2650.0   49853.0     0.0     0.0     0.0     0.0   
166252   720.0  1890.0  ...   760.0  139792.0   170.0   198.0   490.0  3842.0   

        N11901   A11901  N11902   A11902  
0        120.0     94.0  1290.0   2792.0  
1        210.0    301.0  1130.0   2935.0  
2        200.0    376.0   760.0   2058.0  
3        150.0    374.0   460.0   1390.0  
4        220.0    854.0   410.0   1709.0  
...        ...      ...     ...      ...  
166248   550.0    869.0  2960.0   7717.0  
166249   560.0   1273.0  1890.0   5330.0  
166250   470.0   1635.0  1390.0   4696.0  
166251   950.0   5576.0  1460.0   6853.0  
166252   340.0  14487.0   160.0  21167.0  

[166253 rows x 91 columns]

In [19]:
def compute_percentile_zip(df_zip, percentile=0.65):
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

* Compute the 65th income percentile ( percentile=0.65) for each zipcode in `tax_data_valid_zip_cleaned` data frame

  * Your answer can only use `DataFrame` or `Series` methods or properties
  
  * Recall that you can use the `apply` function to transform the values of a column
  
  
* Sort the values in descending order and assign them to a new `DataFrame` called `zip_rev_all`
* The resulting `Series` should look like the following( '...' represents remaining data that is not shown)

```Python
ZIPCODE
33109    3954.114286
33480    3413.301538
...
```

In [20]:
## WRITE YOUR CODE HERE
grouper = tax_data_valid_zip_cleaned.groupby('ZIPCODE').apply(compute_percentile_zip)

In [21]:
zip_rev_all = grouper.sort_values(ascending=False)
zip_rev_all.head()

ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
94027.0    3091.537013
10577.0    2414.855556
dtype: float64

- What are the three zip codes with the most significant 65th percentile value for income?

In [22]:
## WRITE YOUR CODE HERE
zip_rev_all.head(3)

ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
dtype: float64

# 2 Working with the Medicaid Data

### Loading and exploring the data 

* Load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`. The file is located in the `data` directory of the assignment folder. 
* Note that this is quite large and may take some time to load on a computer with modest RAM resources (4GB or less)


In [23]:
## WRITE YOUR CODE HERE
medicaid_data = pd.read_csv('data/medicaid_data.csv')
medicaid_data.head()

Utilization Type State          NDC Product Name  Units Reimbursed  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   Number of Prescriptions  Total Amount Reimbursed  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   Medicaid Amount Reimbursed  Non Medicaid Amount Reimbursed  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               Location  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)


- Modify `medicaid_data` to uppercase all the column names 

  - If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties


In [24]:
## WRITE YOUR CODE HERE
medicaid_data.columns = medicaid_data.columns.str.upper()
medicaid_data.head()

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               LOCATION  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)


- Familiarize your self with the data
  - the `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States
  
  - The remaining column names are self-explanatory
  
- Explore the number of lines and columns in the data

- Check that your column headers are in uppercase

In [25]:
## WRITE YOUR CODE HERE
print(medicaid_data.columns)
print(medicaid_data.shape)
print(medicaid_data.size)

Index(['UTILIZATION TYPE', 'STATE', 'NDC', 'PRODUCT NAME', 'UNITS REIMBURSED',
       'NUMBER OF PRESCRIPTIONS', 'TOTAL AMOUNT REIMBURSED',
       'MEDICAID AMOUNT REIMBURSED', 'NON MEDICAID AMOUNT REIMBURSED',
       'LOCATION'],
      dtype='object')
(1695546, 10)
16955460


* If you explore the  `Location` column for all the entries for which "STATE" value is equal to "HI" you'll notice that all the values are identical

* Are there any states that have more than one value for `Location`. 
  * Hint: think about using a sorted aggregation as part of a split-apply-combine operation to answer this question
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [43]:
hawaii = medicaid_data.groupby('STATE').get_group('HI')
hawaii

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
154                 MCOU    HI  16714034804   DASETTA 1-             448.0   
294                 MCOU    HI  49348002972    SM TRIPLE             821.6   
318                 MCOU    HI    378043301    BUPROPION            5461.0   
397                 MCOU    HI  49348004537   SM ALLERGY            5016.0   
557                 MCOU    HI     54007928   BALSALAZID            3630.0   
...                  ...   ...          ...          ...               ...   
1694941             MCOU    HI  45802006436   TRIAMCINOL           11520.0   
1695124             MCOU    HI    299590825   EPIDUO 0.1            1890.0   
1695291             MCOU    HI  69097084507   CYCLOBENZA            4811.0   
1695327             MCOU    HI  52544024928   MICROGESTI            9576.0   
1695509             MCOU    HI     93227534    AMOX-CLAV            3493.0   

         NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
154                         16.0                   267.07   
294                         27.0                   143.34   
318                         97.0                  2206.56   
397                         35.0                   105.57   
557                         14.0                  3204.75   
...                          ...                      ...   
1694941                    138.0                  1331.15   
1695124                     37.0                  6509.69   
1695291                    129.0                   714.83   
1695327                    131.0                  5166.67   
1695509                    185.0                  2287.13   

         MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
154                          267.07                            0.00   
294                          143.34                            0.00   
318                         2206.56                            0.00   
397                          105.57                            0.00   
557                         3204.75                            0.00   
...                             ...                             ...   
1694941                     1331.15                            0.00   
1695124                     6509.69                            0.00   
1695291                      707.49                            7.34   
1695327                     5166.67                            0.00   
1695509                     2282.69                            4.44   

                     LOCATION  
154      (21.1098, -157.5311)  
294      (21.1098, -157.5311)  
318      (21.1098, -157.5311)  
397      (21.1098, -157.5311)  
557      (21.1098, -157.5311)  
...                       ...  
1694941  (21.1098, -157.5311)  
1695124  (21.1098, -157.5311)  
1695291  (21.1098, -157.5311)  
1695327  (21.1098, -157.5311)  
1695509  (21.1098, -157.5311)  

[13438 rows x 10 columns]

In [26]:
def loc_comp(df):
    df['match'] = df.LOCATION.eq(df.LOCATION.shift())
    return df

In [27]:
groups = medicaid_data.groupby('STATE').apply(loc_comp)

In [42]:
new_df = groups[groups['match'] == False]
counts = new_df.value_counts(new_df['STATE'])
counts

STATE
XX    153253
AL         1
ND         1
NE         1
NH         1
NJ         1
NM         1
NV         1
NY         1
OH         1
OK         1
OR         1
PA         1
RI         1
SC         1
SD         1
TN         1
TX         1
UT         1
VA         1
VT         1
WA         1
WI         1
WV         1
WY         1
NC         1
AK         1
MS         1
IA         1
AR         1
AZ         1
CA         1
CO         1
CT         1
DC         1
DE         1
FL         1
GA         1
HI         1
ID         1
MO         1
IL         1
IN         1
KS         1
KY         1
LA         1
MA         1
MD         1
ME         1
MI         1
MN         1
MT         1
dtype: int64



* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.
  
* The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescription in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).
  
https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained here:

    https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html
    
    
* A parsed/processed version (medicaid_enrollment.tsv) can be in data director of the assignment folder. Use `pandas` to load the medicaid_enrollmen file into DataFrame named `medicaid_enrollment`

In [45]:
## WRITE YOUR CODE HERE
medicaid_enrollment = pd.read_csv('data/medicaid_enrollment.tsv', sep='\t')
medicaid_enrollment

STATE Total Medicaid Enrollees
0                    Alabama                1,050,989
1                     Alaska                  164,783
2             American Samoa                      NaN
3                    Arizona                1,740,520
4                   Arkansas                  762,166
5                 California               13,096,861
6                   Colorado                1,264,600
7                Connecticut                  746,119
8                   Delaware                  227,909
9       District of Columbia                  271,428
10                   Florida                3,808,334
11                   Georgia                1,990,810
12                      Guam                      n/a
13                    Hawaii                  340,513
14                     Idaho                  283,355
15                  Illinois                3,269,999
16                   Indiana                1,295,358
17                      Iowa                  618,505
18                    Kansas                  403,844
19                  Kentucky                1,284,193
20                 Louisiana                1,402,212
21                     Maine                  288,324
22                  Maryland                1,271,445
23             Massachusetts                1,829,618
24                  Michigan                3,947,031
25                 Minnesota                1,052,521
26               Mississippi                  740,937
27                  Missouri                  944,257
28                   Montana                  139,950
29                  Nebraska                  239,463
30                    Nevada                  588,304
31             New Hampshire                  186,399
32                New Jersey                1,705,594
33                New Mexico                  826,155
34                  New York                6,281,038
35            North Carolina                1,965,805
36              North Dakota                   86,250
37  Northern Mariana Islands                      NaN
38                      Ohio                3,060,446
39                  Oklahoma                  829,561
40                    Oregon                1,123,913
41              Pennsylvania                2,569,232
42               Puerto Rico                1,458,819
43              Rhode Island                  308,521
44            South Carolina                1,233,430
45              South Dakota                  124,497
46                 Tennessee                1,562,745
47                     Texas                4,273,982
48                      Utah                  293,867
49                   Vermont                  206,469
50            Virgin Islands                      NaN
51                  Virginia                1,092,225
52                Washington                1,771,679
53             West Virginia                  545,748
54                 Wisconsin                1,209,714
55                   Wyoming                   66,532

* Modify `medicaid_enrollment` to uppercase all the column names 

  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself


In [46]:
## WRITE YOUR CODE HERE
medicaid_enrollment.columns = medicaid_enrollment.columns.str.upper()
medicaid_enrollment.columns

Index(['STATE', 'TOTAL MEDICAID ENROLLEES'], dtype='object')

* Note that some states/territories have missing values. Remove the missing values and save the resulting `DataFrame` as a new variable named `medicaid_enrollment_cleaned`

* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing values means in this context

In [47]:
## WRITE YOUR CODE HERE
medicaid_enrollment_cleaned = medicaid_enrollment.dropna()
medicaid_enrollment_cleaned

STATE TOTAL MEDICAID ENROLLEES
0                Alabama                1,050,989
1                 Alaska                  164,783
3                Arizona                1,740,520
4               Arkansas                  762,166
5             California               13,096,861
6               Colorado                1,264,600
7            Connecticut                  746,119
8               Delaware                  227,909
9   District of Columbia                  271,428
10               Florida                3,808,334
11               Georgia                1,990,810
12                  Guam                      n/a
13                Hawaii                  340,513
14                 Idaho                  283,355
15              Illinois                3,269,999
16               Indiana                1,295,358
17                  Iowa                  618,505
18                Kansas                  403,844
19              Kentucky                1,284,193
20             Louisiana                1,402,212
21                 Maine                  288,324
22              Maryland                1,271,445
23         Massachusetts                1,829,618
24              Michigan                3,947,031
25             Minnesota                1,052,521
26           Mississippi                  740,937
27              Missouri                  944,257
28               Montana                  139,950
29              Nebraska                  239,463
30                Nevada                  588,304
31         New Hampshire                  186,399
32            New Jersey                1,705,594
33            New Mexico                  826,155
34              New York                6,281,038
35        North Carolina                1,965,805
36          North Dakota                   86,250
38                  Ohio                3,060,446
39              Oklahoma                  829,561
40                Oregon                1,123,913
41          Pennsylvania                2,569,232
42           Puerto Rico                1,458,819
43          Rhode Island                  308,521
44        South Carolina                1,233,430
45          South Dakota                  124,497
46             Tennessee                1,562,745
47                 Texas                4,273,982
48                  Utah                  293,867
49               Vermont                  206,469
51              Virginia                1,092,225
52            Washington                1,771,679
53         West Virginia                  545,748
54             Wisconsin                1,209,714
55               Wyoming                   66,532

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

* Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization

 

* Inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and  make sure that the data type is `int`


In [33]:
## WRITE YOUR CODE HERE 

### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

- We can use the shared State information across both tables to associate both tables (SQL JOIN).
- However,  `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name
  - We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`

- Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a Wikipedia page.
  - A brief description of what the code does is included in the comments

In [34]:
# reads all the tables at the given url
# Header=0 instructs pandas to use line 0 as the header (columns)
tables = pd.read_html('https://www.50states.com/abbreviations.htm', header=0)



# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]
Codes_abbreviations.head(5)

HTTPError: HTTP Error 403: Forbidden

* Change the the `DataFrame`'s headers from ['US State:', 'Abbreviation:'] to ['US STATE', 'ABBREVIATION']

  * You can hard code this operation


In [ ]:
## WRITE YOUR CODE HERE 


* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations` 
* Save the results to variable named `medicaid_enrollment_cleaned_with_zip`

- `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
   STATE    Total Medicaid Enrollees    ABBREVIATION
0    Alabama    1,050,989    AL
1    Alaska    164,783    AK
2    Arizona    1,740,520    AZ
3    Arkansas    762,166    AR
4    California    13,096,861    CA
...
```

* We did not cover joins in class -- you find a plethora of examples on how to do this online. See for instance:

`https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html`

* If you cannot get it to work, contact the `TA` for the solution. You will not be penalized if you don't answer this question. 

In [ ]:
## WRITE YOUR CODE HERE 


* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`
  * Remove the column make sure your data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( `...` represents remaining data that is not shown):

```
    Total Medicaid Enrollees    ABBREVIATION
0   1,050,989                  AL
1   164,783                    AK
2   1,740,520                  AZ
3   762,166                    AR
4   13,096,861                 CA
....
```

In [ ]:
## WRITE YOUR CODE HERE 


- Use `DataFrame medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`

   I.E., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the STATE value in the entry.
   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`
- The resulting DataFrame should look like the following (`...` represents remaining data that is not shown):

```
UTILIZATION TYPE    STATE    NDC    PRODUCT NAME    UNITS REIMBURSED    NUMBER OF PRESCRIPTIONS    TOTAL AMOUNT REIMBURSED    MEDICAID AMOUNT REIMBURSED    NON MEDICAID AMOUNT REIMBURSED    LOCATION
0    MCOU    PA    55150023930    Dexamethas    33.0    19.0    234.98    234.98    0.0    (40.5773, -77.264)
1    FFSU    NY    23917710    ALPHAGAN P    570.0    57.0    16006.34    16006.34    0.0    (42.1497, -74.9384)
2    MCOU    OR    13925050501    Dapsone 10    456.0    15.0    1052.42    1052.42    0.0    (44.5672, -122.1269)
...
```

* The order of the columns in the `DataFrame` is not important. This answer uses the same approach as the one used to `merge` the tables above. 

In [ ]:
## WRITE YOUR CODE HERE 


- Remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [ ]:
## WRITE YOUR CODE HERE 



- Use ["STATE", "PRODUCT NAME"] as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns)
  * Hint: the function to set an index on a `DataFrame` can take a single column name or a list of column names. The list here is  ["STATE", "NDC"]
- Call the new data `medicaid_data_w_enrollments_hierarch`
- Inspect your data to make sure the new index has now two levels (STATE and NDC)

In [ ]:
## WRITE YOUR CODE HERE 



* Write a single Pandas expression to print all the lines with containing NDC 61958180101 in "PA"

 * Use a single indexing call (bracket notation) using `loc`
 
 * Hint 1: Since your index is hierarchical, `loc` is expecting two values, the first for STATE and the second for NDC


In [ ]:
## WRITE YOUR CODE HERE 


#### Computing the normalized `UNITS REIMBURSED` per druc 

* Compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.

- For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612

```python
total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum() 
```

- And the numebr of Medicaid Enrollees in "PA" is 2,569,232


```python
total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]
```
- Therefore, the UNITS REIMBURSED per enrollee  for "HARVONI" is  0.00413041718303


```python
print(total_amount_reimbursed/total_enrollees_AK)
```

- Rather than work directly with the ratios, we are going to compute and work with thier logarithm (np.log2) instead.

  - The reason we use logs here is to avoid working small numbers. More on log-transformation in future lectures

- Save the result in a `Series` called `medicaid_reimbursement_per_enrollee`


- Your final result should be a Series that look like the following ( `...` represents data that is not shown here):
  
```
STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
...

AL     2143380    -9.940595
       2143480    -9.805485
       2322730    -5.336260
...

MA     2143380    -7.921997
       2143480    -7.803463
       2144511   -13.472194
       2197590    -7.741402
       2322730    -5.414724
       2322830    -4.592154
       2322930    -4.205626

...


```


In [ ]:
## WRITE YOUR CODE HERE 



- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`

- Using `medicaid_reimbursement_per_enrollee`, generate a `DataFrame` where: 
  - index should be the two-letter state symbol 
  - the column names should be the NDC codes 

- The `DataFrame`  should be formatted as in the image below
  - Hint, simply unstack the data


<img src="media/unstacked.png" alt="drawing" style="width:900px;"/>


In [ ]:
## WRITE YOUR CODE HERE 


#### Exploring the data (very briefly) 
- What is the drug with the highest log-normalized ratio in Hawaii?


In [ ]:
## WRITE YOUR CODE HERE 


* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii 
  * What is it used for?

* Compare the value of `Units Reimbursed` that product between HI and other states, (take for instance MA, FL, OR and WA)
* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii?

In [ ]:
## WRITE YOUR CODE HERE 

In [ ]:
## WRITE YOUR ANSWER HERE 

* Find and list all unique `NDC`s for which the difference between the largest and second large log-normalized ratio by state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio  of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ration of  `-12.623428`
  


In [ ]:
## WRITE YOUR CODE HERE 


- The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states.
   - Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   - Second highest state has a log value of -12.623428 (0.000158478197834722)
- Oklahoma is not a high-incidence state for cancer
- Could the following explain what is happening in Oklahoma?

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10